### ISSUE TITLE
FreeWassersteinBarycenter weights computation

### ISSUE THREAD

**Issue by pamikem** [ott-jax/ott#657]
Hi!

I'm currently playing with the `FreeWassersteinBarycenter `. I noticed that with the output state `FreeBarycenterState`, I have access to the barycenter points `x` and barycenter weights `a`.

After testing on some synthetic data and real-world data, I always had uniform weights `a`. So, I found to way to track the values of the weights through iterations. The values were static (no matter the input data) so I came to the conclusion that the **barycenter weights `a` were not updated**. That would explain why in the initialization, we can only define the initial barycenter points. 

Please, tell me if I'm correct or wrong about the weights `a` not being updated. 

Thank you in advance!

**Comment by marcocuturi** [comment]
Hi @pamikem , thanks for your comment. Indeed, there's a 

`a = jnp.ones((bar_size,)) / bar_size`

statement (with a TODO for me right above ;)) that sets this as uniform. 

Ideally, if I had implemented as advocated in https://arxiv.org/abs/1310.4375 Algorithm 2, I would be re-updating the weights every now and then. I haven't coded this option. This would be akin to running a `DiscreteWassersteinBarycenter` at each (or less frequently) iteration of the `FreeWassersteinBarycenter` update of centroids.

**Comment by pamikem** [comment]
Ok and I guess running `DiscreteWassersteinBarycenter` at each step is equivalent to running Algo 1 at each step of Algo 2 ? (As mentionned in the article).

I'm curious. Did you implement uniform weights `a` by default just because the overall algorithm is less heavier computationnaly ? Or also because the updates of `a`  have no much impact on the convergence of barycenter points `x` ?

**Comment by marcocuturi** [comment]
you have an interesting point there.
- if you use the algorithm from 2014 and let `a` be completely freely reset after each iteration, then you get something like a generalization of the M step in EM for k-means. For instance if the number of measures is 1, then picking the best `a` is actually closed form (you just assign the mass of each point in the measure to the nearest point in the `FreeBarycenter` and count in the end how many were assigned. For 2 measures or more it would be a bit more involved.
- in the paper, my idea was rather to add a constraint or a normalization on `a`
- at the moment I indeed freeze `a`. This is more akin (again, when # measures =1) to the k-means algorithm with a constraint that each centroid has equal mass, i.e. a "well distributed" k-means partition. This is the type of constraint that people have used e.g. for MOE routers https://arxiv.org/pdf/2202.01169 or in SWAV https://arxiv.org/abs/2006.09882

**Comment by pamikem** [comment]
Thank you! I think I understand a bit better how the algorithm works.

Your idea of adding a constraint on `a` is probably justified in the continuous setting. For example, let's suppose we want to compute the barycenter of continuous probability measures. In this case, we can just work with empirical measures when we have enough observations from each input measure. So, I think at some point, it's just enough to consider that the input measures weights `b_i` and the barycenter weights `a` are uniform.

I noticed all this by doing experiments with more than 1000 points per measure.
